File: patch_api.py
import re

filepath = r'C:\Users\Mina Wael\Desktop\CEO\apps\api\src\index.ts'
with open(filepath, 'r', encoding='utf-8') as f:
    content = f.read()

# Add imports if missing
if 'import paymentsRouter from' not in content:
    content = re.sub(
        r"import adminRouter.*?;",
        "import adminRouter, { registerLockdownHandlers } from './routes/admin';\nimport paymentsRouter from './routes/payments';\nimport { requireCredits, requireSubscription } from './middleware';\nimport { deductCredits, CREDIT_COSTS, getUserCredits } from './services/creditEngine';",
        content,
        flags=re.DOTALL
    )

if "app.use('/api/payments', paymentsRouter);" not in content:
    content = re.sub(
        r"app.use\('/api/admin', adminRouter\);",
        "app.use('/api/admin', adminRouter);\napp.use('/api/payments', paymentsRouter);",
        content
    )

# 1. Founder Analysis
content = re.sub(
    r"(app\.post\('/api/founder/analyze', authMiddleware, )(async \(req: Request, res: Response\))",
    r"\1requireCredits(CREDIT_COSTS.FOUNDER_ANALYSIS), \2",
    content
)
content = re.sub(
    r"(const founderProfile = new FounderProfileModel\(\{.*?\}\);\s*if \(dbConnected\) \{)",
    r"await deductCredits(userId, CREDIT_COSTS.FOUNDER_ANALYSIS, 'Founder Analysis');\n    \1",
    content,
    flags=re.DOTALL
)

# 2. Opportunity Discovery
content = re.sub(
    r"(app\.post\('/api/opportunities/discover', authMiddleware, )(async \(req: Request, res: Response\))",
    r"\1requireCredits(CREDIT_COSTS.OPPORTUNITY_DISCOVERY), \2",
    content
)
content = re.sub(
    r"(const formattedOpportunities = \(rawOpportunities \|\| \[\]\))",
    r"await deductCredits(userId, CREDIT_COSTS.OPPORTUNITY_DISCOVERY, 'Opportunity Discovery');\n    \1",
    content
)

# 3. Business Plan
content = re.sub(
    r"(app\.post\('/api/business-plan/generate', authMiddleware, )(async \(req: Request, res: Response\))",
    r"\1requireCredits(CREDIT_COSTS.BUSINESS_PLAN), \2",
    content
)
content = re.sub(
    r"(const businessPlan = new BusinessPlanModel\(\{.*?\}\);\s*if \(dbConnected\) \{)",
    r"await deductCredits(userId, CREDIT_COSTS.BUSINESS_PLAN, 'Business Plan');\n    \1",
    content,
    flags=re.DOTALL
)

# 4. Cofounder Chat
content = re.sub(
    r"(app\.post\('/api/chat/send', authMiddleware, )(async \(req: Request, res: Response\))",
    r"\1requireCredits(CREDIT_COSTS.AI_CHAT_MESSAGE), \2",
    content
)
content = re.sub(
    r"(const userMessage = \{ role: 'user', content: message \};)",
    r"await deductCredits(userId, CREDIT_COSTS.AI_CHAT_MESSAGE, 'AI Chat Message');\n    \1",
    content
)

# Also need an endpoint to get the user's credits
if "app.get('/api/user/credits'" not in content:
    credit_endpoint = """
// Get user credits
app.get('/api/user/credits', authMiddleware, async (req: Request, res: Response): Promise<any> => {
  try {
    const userId = (req as any).user.id;
    const wallet = await getUserCredits(userId);
    return res.status(200).json({ wallet });
  } catch (err: any) {
    return res.status(500).json({ error: err.message });
  }
});
"""
    content = content.replace("// AUTH ROUTES", credit_endpoint + "\n// AUTH ROUTES")

with open(filepath, 'w', encoding='utf-8') as f:
    f.write(content)
print("Patched index.ts")

File: patch-api.js
const fs = require('fs');

let code = fs.readFileSync('apps/api/src/index.ts', 'utf8');

// 1. Imports
code = code.replace(
  `buildContextString,
  getProjectContext
} from '@creator/database';`,
  `buildContextString,
  getProjectContext,
  AIEvaluationModel
} from '@creator/database';`
);

code = code.replace(
  `runPitchAgent } from '@creator/agents';`,
  `runPitchAgent, runEvaluatorAgent } from '@creator/agents';`
);

// 2. Helper
const helperCode = `
// AI Evaluation Helper
async function evaluateAndSave(
  userId,
  projectId,
  targetType,
  targetId,
  input,
  generatedOutput
) {
  if (!dbConnected) return null;
  try {
    const context = await getProjectContext(projectId, userId);
    const { buildEgyptContextString } = await import('@creator/agents');
    const egyptCtx = await buildEgyptContextString(JSON.stringify(input));
    
    const evaluation = await runEvaluatorAgent({
      ventureContext: context,
      egyptMarketContext: egyptCtx,
      generatedOutput,
      targetType
    });
    
    if (evaluation) {
      evaluation.id = \`eval_\${Date.now()}\`;
      evaluation.userId = userId;
      evaluation.projectId = projectId;
      evaluation.targetType = targetType;
      evaluation.targetId = targetId;
      
      const evalDoc = new AIEvaluationModel(evaluation);
      await evalDoc.save();
      return evalDoc.toObject();
    }
  } catch(e) {
    console.error('Evaluation failed:', e);
  }
  return null;
}

// Helper to track Agent Runs`;

code = code.replace('// Helper to track Agent Runs', helperCode);

// 3. Founder Profile Endpoint
code = code.replace(
  /return res\.status\(201\)\.json\(\{ founderProfile \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'founder_profile', founderProfile.id, sanitizedData, founderProfile.toObject());\n    return res.status(201).json({ founderProfile, evaluation });`
);

// 4. Opportunity Discovery
code = code.replace(
  /return res\.json\(\{ opportunities: formattedOpportunities \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'opportunity', formattedOpportunities[0]?.id, founderProfile.toObject(), formattedOpportunities);\n    return res.json({ opportunities: formattedOpportunities, evaluation });`
);

// 5. Business Plan
code = code.replace(
  /return res\.status\(201\)\.json\(\{ businessPlan \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'business_plan', businessPlan.id, selectedOpportunity, businessPlan.toObject());\n    return res.status(201).json({ businessPlan, evaluation });`
);

// 6. Financial Forecast
code = code.replace(
  /return res\.status\(201\)\.json\(\{ financialForecast \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'financial_plan', financialForecast.id, businessPlan, financialForecast.toObject());\n    return res.status(201).json({ financialForecast, evaluation });`
);

// 7. Branding
code = code.replace(
  /return res\.status\(201\)\.json\(\{ brandIdentity \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'branding', brandIdentity.id, businessPlan, brandIdentity.toObject());\n    return res.status(201).json({ brandIdentity, evaluation });`
);

// 8. Marketing
code = code.replace(
  /return res\.status\(201\)\.json\(\{ marketingCampaign \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'marketing', marketingCampaign.id, businessPlan, marketingCampaign.toObject());\n    return res.status(201).json({ marketingCampaign, evaluation });`
);

// 9. Pitch Deck
code = code.replace(
  /return res\.status\(201\)\.json\(\{ pitchDeck \}\);/g,
  `const evaluation = await evaluateAndSave(userId, projectId, 'pitch', pitchDeck.id, businessPlan, pitchDeck.toObject());\n    return res.status(201).json({ pitchDeck, evaluation });`
);


fs.writeFileSync('apps/api/src/index.ts', code);
console.log('Successfully patched index.ts');
